File: dungeons-1.21.11/src/main/java/fishmod/utils/events/CustomEvents.java
package fishmod.utils.events;

import fishmod.utils.debug.Debug;
import fishmod.utils.Location;
import net.fabricmc.fabric.api.client.message.v1.ClientReceiveMessageEvents;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class CustomEvents {

    private static final Pattern PARTY_PATTERN = Pattern.compile("^§9Party §8>");
    private static final Pattern LEAP_PATTERN = Pattern.compile("^You have teleported to .*!$");

    private static final int PARTY_MSG_OFFSET = 11;

    public static void init() {
        //party event
        ClientReceiveMessageEvents.GAME.register((message, overlay) -> {
            String string = message.getString();

            Matcher matcher = PARTY_PATTERN.matcher(string);
            if (!matcher.find()) return;

            int index = string.indexOf(":");
            if (index < PARTY_MSG_OFFSET) {
                Debug.LOGGER.error("{} had bad index", string);
                return;
            }

            String tempUsername = string.substring(PARTY_MSG_OFFSET, index).replaceAll("§.", "");
            if (index + 2 >= string.length()) return;
            String sentMessage = string.substring(index + 2).strip();

            index = tempUsername.indexOf("]") + 2;
            if (index > -1 && index < tempUsername.length()) {
                tempUsername = tempUsername.substring(index);
            }

            String username = tempUsername;
            Events.ON_PARTY_MESSAGE.invoke(partyMessageEvent -> partyMessageEvent.sentMessage(username, sentMessage));
        });

        //leap event
        ClientReceiveMessageEvents.GAME.register((message, overlay) -> {
            if (Location.inDungeon()) {
                String string = message.getString();

                Matcher matcher = LEAP_PATTERN.matcher(string);
                if (!matcher.find()) return;

                Events.ON_LEAP.invoke(leapEvent -> leapEvent.onLeap(message));
            }
        });
    }
}

File: dungeons-26.1.2/src/main/kotlin/fishmod/utils/events/CustomEvents.kt
package fishmod.utils.events

import fishmod.utils.Location
import fishmod.utils.debug.Debug
import net.fabricmc.fabric.api.client.message.v1.ClientReceiveMessageEvents
import java.util.regex.Pattern

object CustomEvents {

    private val PARTY_PATTERN: Pattern = Pattern.compile("^§9Party §8>")
    private val LEAP_PATTERN: Pattern = Pattern.compile("^You have teleported to .*!$")

    private const val PARTY_MSG_OFFSET = 11

    @JvmStatic
    fun init() {
        // party event
        ClientReceiveMessageEvents.GAME.register { message, _ ->
            val string = message.string

            val matcher = PARTY_PATTERN.matcher(string)
            if (!matcher.find()) return@register

            var index = string.indexOf(":")
            if (index < PARTY_MSG_OFFSET) {
                Debug.LOGGER.error("{} had bad index", string)
                return@register
            }

            var tempUsername = string.substring(PARTY_MSG_OFFSET, index).replace(Regex("§."), "")
            if (index + 2 >= string.length) return@register
            val sentMessage = string.substring(index + 2).trim()

            index = tempUsername.indexOf("]") + 2
            if (index > -1 && index < tempUsername.length) {
                tempUsername = tempUsername.substring(index)
            }

            val username = tempUsername
            Events.ON_PARTY_MESSAGE.invoke { partyMessageEvent -> partyMessageEvent.sentMessage(username, sentMessage) }
        }

        // leap event
        ClientReceiveMessageEvents.GAME.register { message, _ ->
            if (Location.inDungeon()) {
                val string = message.string

                val matcher = LEAP_PATTERN.matcher(string)
                if (!matcher.find()) return@register

                Events.ON_LEAP.invoke { leapEvent -> leapEvent.onLeap(message) }
            }
        }
    }
}
